chore: integrate peerdas-kzg

kevaundray opened the pull request:
**Motivation**

This is a Draft PR to experiment with integrating the rust peerdas-kzg library alongside c-kzg. 

Currently playing around with the API to understand how things are integrated into lodestar.

Current notes:

- The synchronous methods are used for backwards compatibility but async methods are available.

**Description**

<!-- A clear and concise general description of the changes of this PR commits -->

<!-- If applicable, add screenshots to help explain your solution -->

<!-- Link to issues: Resolves #111, Resolves #222 -->

Closes #issue_number

**Steps to test or reproduce**

<!--Steps to reproduce the behavior:
```sh
git checkout <feature_branch>
lodestar beacon --new-flag option1
```
-->

g11tech committed:
feat: placeholder PR for electra

add types stub and epoch config

fix types

g11tech committed:
feat: placeholder PR for electra

add types stub and epoch config

fix types

g11tech committed:
feat: implement peerDAS on electra

add some presets

add further params and types

add data column to types repo and network

move to max request data columns to preset

add the datacolumns data in blockinput and fix breaking errors in seen gossip blockinput

handle data columns in gossip and the seengossip

further propagate forkaware blockdata and resolve build/type issues

further handle datacolumns sync by range by root and forkaware data handling

fix issues

chore: update c-kzg to peerDas version

feat: add peerDas ckzg functions to interface

fix the lookups

handle the publishing flow

various sync try fixes

fixes

compute blob side car

various misl debuggings and fixes

debug and apply fixes and get range and by root sync to work will full custody

enable syncing with lower custody requirement

use node peerid rather than a dummy string

get and use the nodeid from enr and correctly compute subnets and column indexes

filterout and connect to peers only matching out custody requiremnt

try adding custody requirement

add protection for subnet calc

get the sync working with devnet 0

correctly set the enr with custody subnet info

rebase fixes

small refactor

g11tech committed:
feat: implement peerDAS on electra

add some presets

add further params and types

add data column to types repo and network

move to max request data columns to preset

add the datacolumns data in blockinput and fix breaking errors in seen gossip blockinput

handle data columns in gossip and the seengossip

further propagate forkaware blockdata and resolve build/type issues

further handle datacolumns sync by range by root and forkaware data handling

fix issues

chore: update c-kzg to peerDas version

feat: add peerDas ckzg functions to interface

fix the lookups

handle the publishing flow

various sync try fixes

fixes

compute blob side car

various misl debuggings and fixes

debug and apply fixes and get range and by root sync to work will full custody

enable syncing with lower custody requirement

use node peerid rather than a dummy string

get and use the nodeid from enr and correctly compute subnets and column indexes

filterout and connect to peers only matching out custody requiremnt

try adding custody requirement

add protection for subnet calc

get the sync working with devnet 0

correctly set the enr with custody subnet info

rebase fixes

small refactor

matthewkeil committed:
fix: docker build issue for c-kzg

wip: REPLACE THIS COMMIT

commit yarn lock

rebase fixes

fix: update c-zkg install workflow

feat: add trustedSetupPrecompute cli flag

fix: update trusted-setup for testing

fix: update c-zkg install workflow to remove sudo

fix: add rsync to apk deps

matthewkeil committed:
fix: docker build issue for c-kzg

wip: REPLACE THIS COMMIT

commit yarn lock

g11tech committed:
rebase fixes

matthewkeil committed:
fix: update c-zkg install workflow

matthewkeil committed:
feat: add trustedSetupPrecompute cli flag

matthewkeil committed:
fix: update trusted-setup for testing

matthewkeil committed:
fix: update c-zkg install workflow to remove sudo

kevaundray committed:
add dep

kevaundray committed:
update peerdas lib

kevaundray committed:
add rough idea

kevaundray committed:
modify code to call the exported computeCellsAndProofs method

kevaundray committed:
add false to use ckzg

kevaundray committed:
add todo re lib switching

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ChainSafe/lodestar?pullRequest=6923) <br/>All committers have signed the CLA.

kevaundray committed:
revert yarn.lock

kevaundray committed:
yarn add "@crate-crypto/peerdas-kzg@0.2.2-5f4e791.0" --exact

kevaundray committed:
use tsc --define to set the library to use

g11tech committed:
feat: get various sync mechanisms working with/without sharded data

g11tech committed:
feat: add the modifications to work with devnet2

some network options to control peering behavior

allow setting node custody capability via --params

use eip754 names for the peerdas config

g11tech committed:
fix: refactor to add and use nodeid computation and clear out nodeid tracking

nodeid cleanup for network

g11tech committed:
fix the types/test

rebase fixes

g11tech committed:
feat: add and use metadatav3 for peer custody subnet

fixes for metadata, working locally

change the condition to update metadata csc change

kevaundray committed:
Merge branch 'peerDAS' into kw/peerdas-rust-lib-integration

kevaundray committed:
use v0.4.1 plus name-change

kevaundray committed:
refactor: package name has changed

kevaundray committed:
yarn lint

kevaundray commented on packages/cli/src/cmds/beacon/initPeerIdAndEnr.ts:
```diff
@@ -87,6 +87,7 @@ export function overwriteEnrWithCliArgs(
     maybeUpdateEnr(enr, "tcp", args["enr.tcp"] ?? port ?? enr.tcp);
     maybeUpdateEnr(enr, "tcp6", args["enr.tcp6"] ?? port6 ?? enr.tcp6);
   }
+  enr.set("custody_subnet_count", ssz.UintNum64.serialize(config.CUSTODY_REQUIREMENT));
```
```suggestion
```

kevaundray commented on packages/types/src/types.ts:
```diff
@@ -164,6 +164,40 @@ type TypesByFork = {
     SyncCommittee: altair.SyncCommittee;
     SyncAggregate: altair.SyncAggregate;
   };
+  [ForkName.electra]: {
+    BeaconBlockHeader: phase0.BeaconBlockHeader;
+    SignedBeaconBlockHeader: phase0.SignedBeaconBlockHeader;
+    BeaconBlock: electra.BeaconBlock;
+    BeaconBlockBody: electra.BeaconBlockBody;
+    BeaconState: electra.BeaconState;
+    SignedBeaconBlock: electra.SignedBeaconBlock;
+    Metadata: altair.Metadata;
+    LightClientHeader: electra.LightClientHeader;
+    LightClientBootstrap: electra.LightClientBootstrap;
+    LightClientUpdate: electra.LightClientUpdate;
+    LightClientFinalityUpdate: electra.LightClientFinalityUpdate;
+    LightClientOptimisticUpdate: electra.LightClientOptimisticUpdate;
+    LightClientStore: electra.LightClientStore;
+    BlindedBeaconBlock: electra.BlindedBeaconBlock;
+    BlindedBeaconBlockBody: electra.BlindedBeaconBlockBody;
+    SignedBlindedBeaconBlock: electra.SignedBlindedBeaconBlock;
+    ExecutionPayload: electra.ExecutionPayload;
+    ExecutionPayloadHeader: electra.ExecutionPayloadHeader;
+    BuilderBid: electra.BuilderBid;
+    SignedBuilderBid: electra.SignedBuilderBid;
+    SSEPayloadAttributes: electra.SSEPayloadAttributes;
+    BlockContents: {block: BeaconBlock<ForkName.electra>; kzgProofs: deneb.KZGProofs; blobs: deneb.Blobs};
+    SignedBlockContents: {
+      signedBlock: SignedBeaconBlock<ForkName.electra>;
+      kzgProofs: deneb.KZGProofs;
+      blobs: deneb.Blobs;
+    };
+    ExecutionPayloadAndBlobsBundle: deneb.ExecutionPayloadAndBlobsBundle;
+    BlobsBundle: deneb.BlobsBundle;
+    Contents: deneb.Contents;
+    SyncCommittee: altair.SyncCommittee;
+    SyncAggregate: altair.SyncAggregate;
+  };
```
```suggestion
```

kevaundray commented on packages/beacon-node/src/network/peers/peersData.ts:
```diff
@@ -24,6 +24,7 @@ export type PeerData = {
   agentVersion: string | null;
   agentClient: ClientKind | null;
   encodingPreference: Encoding | null;
+  custodySubnetCount: number | null;
```
```suggestion
```

kevaundray commented on packages/beacon-node/src/network/peers/peerManager.ts:
```diff
@@ -675,6 +675,7 @@ export class PeerManager {
       agentVersion: null,
       agentClient: null,
       encodingPreference: null,
+      custodySubnetCount,
```
```suggestion
```

kevaundray commented on packages/beacon-node/src/network/peers/discover.ts:
```diff
@@ -354,6 +356,11 @@ export class PeerDiscovery {
     }
     // async due to some crypto that's no longer necessary
     const peerId = await enr.peerId();
+
+    const nodeId = fromHexString(enr.nodeId);
+    this.peerIdToNodeId.set(peerId.toString(), nodeId);
+    pruneSetToMax(this.peerIdToNodeId, MAX_CACHED_NODEIDS);
+
```
```suggestion
```

kevaundray commented on packages/beacon-node/src/network/peers/discover.ts:
```diff
@@ -23,6 +23,8 @@ const MAX_CACHED_ENRS = 100;
 /** Max age a cached ENR will be considered for dial */
 const MAX_CACHED_ENR_AGE_MS = 5 * 60 * 1000;
 
+const MAX_CACHED_NODEIDS = 10000;
+
```
```suggestion
```

kevaundray committed:
Fix: github has problems merging branch due to force merge

Manually remove changes that github did not remove

kevaundray committed:
revert lockfile and regenerate by running `yarn`

kevaundray committed:
change import to node-eth-kzg

kevaundray committed:
update API and add notes on how to upgrade

matthewkeil commented:
closed in favor of #7936 #7989
